### ISSUE TITLE
App store purchase dialog not prompting all the time 

### ISSUE THREAD

**Issue by Tom3652** [RevenueCat/purchases-flutter#364]
**Describe the bug**

Hello guys, i am having the exact same issue #36 under iOS 15.4.
My app got rejected today by Apple because of this (they didn't see the prompt). 
When i tried in debug mode i got these logs : 

```
[Purchases] - DEBUG: ℹ️ Requesting products from the store with identifiers: {(
    "premium_subscription"
)}
[Purchases] - DEBUG: ℹ️ Products request finished.
[Purchases] - DEBUG: 💰 Retrieved SKProducts:
[Purchases] - DEBUG: 💰 premium_subscription - <SKProduct: 0x2829962a0>
[Purchases] - DEBUG: ℹ️ 1 completion handlers waiting on products
[Purchases] - DEBUG: ℹ️ makePurchase
[Purchases] - DEBUG: 💰 Purchasing product - premium_subscription
[Purchases] - DEBUG: ℹ️ PaymentQueue updatedTransaction: premium_subscription 2000000046660027 ((null)) 2000000046618823 - 1
[Purchases] - DEBUG: ℹ️ Loaded receipt from url file:///private/var/mobile/Containers/Data/Application/A0C65C0D-77A4-44A0-AFB5-D8FDC588283F/StoreKit/sandboxReceipt
[Purchases] - DEBUG: ℹ️ Found 0 unsynced attributes for App User ID: GyLmrH9mpDaKpRPP9V2rUSfO9oq1
[Purchases] - DEBUG: ℹ️ There are no requests currently running, starting request POST /receipts
[Purchases] - DEBUG: ℹ️ API request started: POST /v1/receipts
8.15.0 - [Firebase/Analytics][I-ACS023141] Purchase is a duplicate and will not be reported. Product ID: premium_subscription.   
[Purchases] - DEBUG: ℹ️ API request completed with status: POST /v1/receipts 200
[Purchases] - DEBUG: ℹ️ Serial request done: POST /receipts, 0 requests left in the queue
[Purchases] - DEBUG: 💰 Finishing transaction premium_subscription 2000000046660027 (2000000046618823)
flutter: Info : {entitlements: {all: {premium: {identifier: premium, isActive: false, willRenew: false, latestPurchaseDate: 2022-05-04T11:41:15Z, originalPurchaseDate: 2022-05-04T10:40:34Z, productIdentifier: premium_subscription, isSandbox: true, ownershipType: PURCHASED, store: APP_STORE, periodType: NORMAL, expirationDate: 2022-05-04T11:46:15Z, unsubscribeDetectedAt: 2022-05-04T11:46:24Z, billingIssueDetectedAt: null}}, active: {}}, allPurchaseDates: {premium_subscription: 2022-05-04T11:41:15Z}, activeSubscriptions: [], allPurchasedProductIdentifiers: [premium_subscription], nonSubscriptionTransactions: [], firstSeen: 2022-05-02T17:54:08Z, originalAppUserId: $RCAnonymousID:2adaea43f45542f987e7f965ab7b1189, allExpirationDates: {premium_subscription: 2022-05-04T11:46:15Z}, requestDate: 2022-05-04T15:48:35Z, latestExpirationDate: 2022-05-04T11:46:15Z, originalPurchaseDate: 2013-08-01T07:00:00Z, originalApplicationVersion: 1.0, managementURL: null}
[Purchases] - DEBUG: ℹ️ PaymentQueue removedTransaction: premium_subscription 2000000046660027 (2000000046618823 (null)) (null) - 1
```

Even a newly RevenueCat created users. 
It was working fine yesterday in debug mode with my app store sandbox account. I have tried again after Apple rejection and also tried to sign out from the sandbox account to get a real case scenario but the prompt didn't appear either. 

I have also tried removing all users from RevenueCat dashboard (to remove all pending transactions etc..) but this didn't seem to change anything unfortunately. 

I am using `flutter: 2.10.5` stable and `purchases_flutter: ^3.10.0`. 
Prompt should appear after this line : 

```
[Purchases] - DEBUG: ℹ️ API request completed with status: POST /v1/receipts 200
```

Yesterday i also got a prompt saying "Subscription already bought" or something, so i know that the prompt should be displayed even if my UI has an issue (should not be shown if the subscription is active). 
Also as you can see, the result of this is that the subscription is anyway not active (because the user didn't pay obviously).

Flutter doctor : 

```
[✓] Flutter (Channel stable, 2.10.5, on macOS 12.3.1 21E258 darwin-x64, locale fr-FR)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.66.2)
[✓] Connected device (3 available)
[✓] HTTP Host Availability

• No issues found!
```

I have tested on 3 different iOS devices and the issue remains. 

**Additional context**

When i made the issue, my app got rejected by Apple because of the prompt not showing.  
Today it was accepted probably because it was working, so i have downloaded the latest version of my app (from the App store) and i have tried myself the subscription on a new account : the prompt was not showing... 

This means that half of the time the prompt is not showing. 
I can't bet on luck to get my users subscribe to my app... 

Thanks for your understanding

**Comment by Tom3652** [comment]
The prompt started to show again after i have removed the sandbox account from my App Store config. 

However i have another issue now that tells me always "SandBox" for Apple when i have no sandbox account. 

Does RevenueCat always show SandBox dialog ? I have built the app in `release` and `profile` mode with my real Apple account (no sandbox users i have deleted them all in App Store connect) and i have anyway the Sandbox prompt showing.

**Comment by joshdholtz** [comment]
@Tom3652 Nothing _really_ stands out that could be causing an issue. One thing I'm curious about is how far your code is going _when_ you experience that issue the prompt not showing up 🤔 

Do you have any logs to know that it's getting into `await Purchases.purchaseProduct`? I can't think of why it wouldn't based on what your `purchase()` method looks like. The log posted in your PR description looks like it was from a successful purchases because there was a receipt that was posted.

This can either fail:
1) Before `await Purchases.purchaseProduct`
2) In `await Purchases.purchaseProduct` but before calling Apple's APIs
3) In Apple's IAP APIs

I want to see if we can eliminate one of two of these options 😇 If you are able to some debug output from the console when this doesn't work showing that its getting into your `purchase()` method we should then see some logs from inside of `await Purchases.purchaseProduct` that should let us know how far its getting 🤞 

Thanks!!

**Comment by Tom3652** [comment]
Hello @joshdholtz,

My bad so far, i have not seen that my subscription was rejected by the app store, i have linked it to my app's review but it seems it was not under review... 
So i have submitted a new review and will see if the problem persist 🙌🏼 
If this still happens afterwards, i will try what you asked !

Sorry for the disturbance and thanks again for your help ! 🚀

**Comment by joshdholtz** [comment]
@Tom3652 No worries at all! We are always here to help no matter what the question or issue is 🥰 Hoping to hear good news from you soon!

**Comment by Tom3652** [comment]
Well the subscription was approved, but still not appearing unfortunately :/ 
I can see this kind of logs from my logging tool for release mode in real time : 

```
PlatformException(5, Couldn't find product., {code: 5, message: Couldn't find product., userCancelled: false, underlyingErrorMessage: }, null)
```

But the product is found in debug mode so i don't understand... 

Here is my item from App Store connect (sorry it's in french) : 
![Capture d’écran 2022-05-17 à 14 35 59](https://user-images.githubusercontent.com/32631467/168812176-499c65c3-33ac-403b-b436-2b0de5679ce5.png)

And in my RevenueCat setup : 

![Capture d’écran 2022-05-17 à 14 36 31](https://user-images.githubusercontent.com/32631467/168812276-c5a7bf96-a32d-45a2-8a1e-9416670bdf7f.png)

(The code in the above comment is the one used for this log)

Do you see anything wrong with this ? :/ 
I actually got a weird bug in App Store connect, my subscription was waiting for review, then i have removed the customized langages for the subscription here that were refused by Apple : 
![Capture d’écran 2022-05-17 à 14 38 44](https://user-images.githubusercontent.com/32631467/168812678-c67d379a-9ce1-43ab-999b-a43c1783a2b3.png)

And the subscription automatically went in `approved` state... So maybe there is a bug with Apple again :(

**Comment by Tom3652** [comment]
Alright it started to work after a new release, it was indeed an Apple issue like so often 😅
Thanks for your time !

**Comment by joshdholtz** [comment]
@Tom3652 Sorry for the slow reply! I've been in and out the past few days but so glad things are working now 🥳 Feel free to reach out to us with any other issues or questions in the future!
